tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1305 @@
+    /// let mut server = NamedPipeOptions::new().create(PIPE_NAME)?;
+    /// let mut client = NamedPipeClientOptions::new().create(PIPE_NAME)?;
+    /// server.connect().await?;
+    ///
+    /// let client = tokio::spawn(async move {
+    ///     for _ in 0..N {
+    ///         client.write_all(b"ping").await?;
+    ///     }
+    ///
+    ///     let mut buf = [0u8; 4];
+    ///     client.read_exact(&mut buf).await?;
+    ///
+    ///     Ok::<_, io::Error>(buf)
+    /// });
+    ///
+    /// let mut buf = [0u8; 4];
+    /// let mut available = 0;
+    ///
+    /// for n in 0..N {
+    ///     if available < 4 {
+    ///         server.read_exact(&mut buf).await?;
+    ///         assert_eq!(&buf[..], b"ping");
+    ///
+    ///         let info = server.peek(None)?;
+    ///         available = info.total_bytes_available;
+    ///         continue;
+    ///     }
+    ///
+    ///     // here we know that at least `available` bytes are immediately
+    ///     // ready to read.
+    ///     server.read_exact(&mut buf).await?;
+    ///     available -= buf.len();
+    ///     assert_eq!(&buf[..], b"ping");
+    /// }
+    ///
+    /// server.write_all(b"pong").await?;
+    ///
+    /// let buf = client.await??;
+    /// assert_eq!(&buf[..], b"pong");
+    /// # Ok(()) }
+    /// ```
+    pub fn peek(&mut self, mut buf: Option<&mut ReadBuf<'_>>) -> io::Result<PipePeekInfo> {
+        use std::convert::TryFrom as _;
+
+        unsafe {
+            let mut n = mem::MaybeUninit::zeroed();
+            let mut total_bytes_available = mem::MaybeUninit::zeroed();
+            let mut bytes_left_this_message = mem::MaybeUninit::zeroed();
+
+            let result = {
+                let (buf, len) = match &mut buf {
+                    Some(buf) => {
+                        let len = DWORD::try_from(buf.capacity())
+                            .expect("buffer too large for win32 api");
+                        // Safety: the OS has no expectation on whether the
+                        // buffer is initialized or not.
+                        let buf = buf.inner_mut() as *mut _ as *mut _;
+                        (buf, len)
+                    }
+                    None => (ptr::null_mut(), 0),
+                };
+
+                namedpipeapi::PeekNamedPipe(
+                    self.io.as_raw_handle(),
+                    buf,
+                    len,
+                    n.as_mut_ptr(),
+                    total_bytes_available.as_mut_ptr(),
+                    bytes_left_this_message.as_mut_ptr(),
+                )
+            };
+
+            if result == FALSE {
+                return Err(io::Error::last_os_error());
+            }
+
+            let n = usize::try_from(n.assume_init()).expect("output size too large");
+
+            if let Some(buf) = buf {
+                // Safety: we trust that the OS has initialized up until `n`
+                // through the call to PeekNamedPipe.
+                buf.assume_init(n);
```

Comment:
You need to also call [`advance`](https://docs.rs/tokio/1/tokio/io/struct.ReadBuf.html#method.advance).
```suggestion
                // Safety: we trust that the OS has initialized up until `n`
                // through the call to PeekNamedPipe.
                buf.assume_init(n);
                buf.advance(n);
```